### PR TITLE
Add metrics for cluster-operator

### PR DESCRIPTION
**Pull request by OchiengEd**
### Type of change
- Enhancement / new feature

### Description
This PR adds metrics to the cluster-operator. At this time, we have primarily have generic metrics; Jvm metrics(garbage collection, memory, etc) and processor metrics. I have included metrics available below.

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [ ] Update/write design documentation in `./design`
- [ ] Write tests
- [ ] Make sure all tests pass
- [ ] Update documentation
- [ ] Check RBAC rights for Kubernetes / OpenShift roles
- [ ] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [ ] Reference relevant issue(s) and close them after merging
- [ ] Update CHANGELOG.md

```
# curl http://10.129.2.97:8080/metrics
# HELP vertx_pool_queue_size Number of elements waiting for a resource
# TYPE vertx_pool_queue_size gauge
vertx_pool_queue_size{pool_type="worker",} 0.0
# HELP vertx_verticle_deployed Number of verticle instances deployed
# TYPE vertx_verticle_deployed gauge
vertx_verticle_deployed 1.0
# HELP jvm_buffer_count_buffers An estimate of the number of buffers in the pool
# TYPE jvm_buffer_count_buffers gauge
jvm_buffer_count_buffers{id="direct",} 1.0
jvm_buffer_count_buffers{id="mapped",} 0.0
# HELP jvm_memory_committed_bytes The amount of memory in bytes that is committed for the Java virtual machine to use
# TYPE jvm_memory_committed_bytes gauge
jvm_memory_committed_bytes{area="heap",id="Tenured Gen",} 1.3033472E7
jvm_memory_committed_bytes{area="heap",id="Eden Space",} 5308416.0
jvm_memory_committed_bytes{area="nonheap",id="Metaspace",} 3.2374784E7
jvm_memory_committed_bytes{area="nonheap",id="Code Cache",} 6619136.0
jvm_memory_committed_bytes{area="heap",id="Survivor Space",} 655360.0
jvm_memory_committed_bytes{area="nonheap",id="Compressed Class Space",} 3801088.0
# HELP jvm_memory_max_bytes The maximum amount of memory in bytes that can be used for memory management
# TYPE jvm_memory_max_bytes gauge
jvm_memory_max_bytes{area="heap",id="Tenured Gen",} 8.8080384E7
jvm_memory_max_bytes{area="heap",id="Eden Space",} 3.5258368E7
jvm_memory_max_bytes{area="nonheap",id="Metaspace",} -1.0
jvm_memory_max_bytes{area="nonheap",id="Code Cache",} 2.5165824E8
jvm_memory_max_bytes{area="heap",id="Survivor Space",} 4390912.0
jvm_memory_max_bytes{area="nonheap",id="Compressed Class Space",} 1.073741824E9
# HELP vertx_http_server_bytesReceived_max Number of bytes received by the server
# TYPE vertx_http_server_bytesReceived_max gauge
vertx_http_server_bytesReceived_max 0.0
# HELP vertx_http_server_bytesReceived Number of bytes received by the server
# TYPE vertx_http_server_bytesReceived summary
vertx_http_server_bytesReceived_count 34.0
vertx_http_server_bytesReceived_sum 0.0
# HELP jvm_buffer_total_capacity_bytes An estimate of the total capacity of the buffers in this pool
# TYPE jvm_buffer_total_capacity_bytes gauge
jvm_buffer_total_capacity_bytes{id="direct",} 0.0
jvm_buffer_total_capacity_bytes{id="mapped",} 0.0
# HELP jvm_classes_unloaded_classes_total The total number of classes unloaded since the Java virtual machine has started execution
# TYPE jvm_classes_unloaded_classes_total counter
jvm_classes_unloaded_classes_total 0.0
# HELP vertx_http_server_responseTime_seconds Request processing time
# TYPE vertx_http_server_responseTime_seconds summary
vertx_http_server_responseTime_seconds_count{code="200",method="HEAD",} 1.0
vertx_http_server_responseTime_seconds_sum{code="200",method="HEAD",} 5.73166E-4
vertx_http_server_responseTime_seconds_count{code="200",method="GET",} 33.0
vertx_http_server_responseTime_seconds_sum{code="200",method="GET",} 0.031361446
# HELP vertx_http_server_responseTime_seconds_max Request processing time
# TYPE vertx_http_server_responseTime_seconds_max gauge
vertx_http_server_responseTime_seconds_max{code="200",method="HEAD",} 5.73166E-4
vertx_http_server_responseTime_seconds_max{code="200",method="GET",} 4.41798E-4
# HELP vertx_http_server_requestCount_total Number of processed requests
# TYPE vertx_http_server_requestCount_total counter
vertx_http_server_requestCount_total{code="200",method="HEAD",} 1.0
vertx_http_server_requestCount_total{code="200",method="GET",} 33.0
# HELP vertx_pool_completed_total Number of elements done with the resource
# TYPE vertx_pool_completed_total counter
vertx_pool_completed_total{pool_type="worker",} 11.0
# HELP vertx_pool_ratio Pool usage ratio, only present if maximum pool size could be determined
# TYPE vertx_pool_ratio gauge
vertx_pool_ratio{pool_type="worker",} 0.0
# HELP jvm_memory_used_bytes The amount of used memory
# TYPE jvm_memory_used_bytes gauge
jvm_memory_used_bytes{area="heap",id="Tenured Gen",} 9346776.0
jvm_memory_used_bytes{area="heap",id="Eden Space",} 3381800.0
jvm_memory_used_bytes{area="nonheap",id="Metaspace",} 3.1488472E7
jvm_memory_used_bytes{area="nonheap",id="Code Cache",} 6555712.0
jvm_memory_used_bytes{area="heap",id="Survivor Space",} 655352.0
jvm_memory_used_bytes{area="nonheap",id="Compressed Class Space",} 3593712.0
# HELP vertx_pool_usage_seconds Time using a resource
# TYPE vertx_pool_usage_seconds summary
vertx_pool_usage_seconds_count{pool_type="worker",} 11.0
vertx_pool_usage_seconds_sum{pool_type="worker",} 0.984596539
# HELP vertx_pool_usage_seconds_max Time using a resource
# TYPE vertx_pool_usage_seconds_max gauge
vertx_pool_usage_seconds_max{pool_type="worker",} 0.0
# HELP vertx_http_server_bytesSent Number of bytes sent by the server
# TYPE vertx_http_server_bytesSent summary
vertx_http_server_bytesSent_count 34.0
vertx_http_server_bytesSent_sum 0.0
# HELP vertx_http_server_bytesSent_max Number of bytes sent by the server
# TYPE vertx_http_server_bytesSent_max gauge
vertx_http_server_bytesSent_max 0.0
# HELP vertx_pool_queue_delay_seconds Queue time for a resource
# TYPE vertx_pool_queue_delay_seconds summary
vertx_pool_queue_delay_seconds_count{pool_type="worker",} 11.0
vertx_pool_queue_delay_seconds_sum{pool_type="worker",} 1.925301221
# HELP vertx_pool_queue_delay_seconds_max Queue time for a resource
# TYPE vertx_pool_queue_delay_seconds_max gauge
vertx_pool_queue_delay_seconds_max{pool_type="worker",} 0.0
# HELP jvm_classes_loaded_classes The number of classes that are currently loaded in the Java virtual machine
# TYPE jvm_classes_loaded_classes gauge
jvm_classes_loaded_classes 5111.0
# HELP vertx_http_server_connections Number of opened connections to the server
# TYPE vertx_http_server_connections gauge
vertx_http_server_connections 1.0
# HELP vertx_pool_inUse Number of resources used
# TYPE vertx_pool_inUse gauge
vertx_pool_inUse{pool_type="worker",} 0.0
# HELP vertx_http_server_requests Number of requests being processed
# TYPE vertx_http_server_requests gauge
vertx_http_server_requests{method="GET",} 1.0
vertx_http_server_requests{method="HEAD",} 0.0
# HELP jvm_buffer_memory_used_bytes An estimate of the memory that the Java virtual machine is using for this buffer pool
# TYPE jvm_buffer_memory_used_bytes gauge
jvm_buffer_memory_used_bytes{id="direct",} 1.0
jvm_buffer_memory_used_bytes{id="mapped",} 0.0
#
```

### CODE REVIEW

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/ClusterOperator.java [hunk 1]
```diff
@@ -171,6 +185,15 @@ private void reconcileAll(String trigger) {
         return result.future();
     }
 
+    private void setupMetrics() {
+        new ClassLoaderMetrics().bindTo(registry);
+        new JvmMemoryMetrics().bindTo(registry);
+        new ProcessorMetrics().bindTo(registry);
+        new JvmThreadMetrics().bindTo(registry);
+        new JvmGcMetrics().bindTo(registry);
+        new KafkaConsumerMetrics().bindTo(registry);
```
Comment (scholzj) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/ClusterOperator.java]: I wonder if the `KafkaConsumerMetrics` have any value given the Cluster operator does not use any actual KafkaConsumer but only an KAfka Admin client. Does it return any metrics?

File: cluster-operator/src/main/resources/kafka-2.2.1-config-model.json [hunk 2]
```diff
@@ -1,825 +0,0 @@
-{
```
Comment (scholzj) [inline cluster-operator/src/main/resources/kafka-2.2.1-config-model.json]: The `cluster-operator/src/main/resources/kafka-*-config-model.json` files are generated and should be part of the code base. I need to have a look wh did it got deleted. Could you please restore them int he PR in the mean time?

@tombentley Why are these actually in the code base? Isn't it enough to have these in the binaries?

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/ClusterOperator.java (hunk 1 above)
Comment (OchiengEd) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/ClusterOperator.java]: I was thinking about this. I have removed the KafkaConsumer metrics.

File: cluster-operator/src/main/resources/kafka-2.2.1-config-model.json (hunk 2 above)
Comment (OchiengEd) [reply cluster-operator/src/main/resources/kafka-2.2.1-config-model.json]: the mentioned files were added back.

File: cluster-operator/pom.xml [hunk 3]
```diff
@@ -222,6 +222,21 @@
...
+        </dependency>
+        <dependency>
+            <groupId>io.micrometer</groupId>
+            <artifactId>micrometer-registry-prometheus</artifactId>
+            <version>1.3.1</version>
+        </dependency>
+        <dependency>
+            <groupId>io.vertx</groupId>
+            <artifactId>vertx-micrometer-metrics</artifactId>
+            <version>3.8.4</version>
```
[the remark this answers is not in the document]
Comment (OchiengEd) [reply cluster-operator/pom.xml]: Resolved. Thanks

File: cluster-operator/pom.xml [hunk 4]
```diff
@@ -222,6 +222,21 @@
...
         </dependency>
+        <dependency>
+            <groupId>io.micrometer</groupId>
+            <artifactId>micrometer-core</artifactId>
+            <version>1.3.1</version>
+        </dependency>
+        <dependency>
+            <groupId>io.micrometer</groupId>
+            <artifactId>micrometer-registry-prometheus</artifactId>
+            <version>1.3.1</version>
```
[the remark this answers is not in the document]
Comment (OchiengEd) [reply cluster-operator/pom.xml]: Resolved. Thanks

**Comment by OchiengEd** [comment]
> The code in the PR looks good to me. However, it seems the tests in the `ClusterOperatorTest` class are failing with NPE. I think the tests there will probably need to mock some of the variables. Could you please have a look at it? (or let me know and I can try to help with it). Thanks

Am reviewing the `ClusterOperatorTest` class. Will let you know if I need assistance.

**Comment by scholzj** [comment]
@ppatierno Do you have anything more?

**Comment by scholzj** [comment]
Thanks for the PR.

**Pull request merged**